src/tvu/repair/service.zig

```diff
@@ -0,0 +1,436 @@
+        const request: RepairRequest = .{ .HighestShred = .{ slot, 0 } };
+        const maybe_peer = try self.peers.getRandomPeer(slot);
+
+        if (maybe_peer) |peer| return .{
+            .request = request,
+            .recipient = peer.pubkey,
+            .recipient_addr = peer.serve_repair_socket,
+        } else {
+            return null;
+        }
+    }
+};
+
+/// Signs and serializes repair requests. Sends them over the network.
+pub const RepairRequester = struct {
+    allocator: Allocator,
+    rng: Random,
+    keypair: *const KeyPair,
+    udp_send_socket: *Socket,
+    logger: Logger,
+
+    pub fn sendRepairRequest(
+        self: *const @This(),
+        request: AddressedRepairRequest,
+    ) !void {
+        const timestamp = std.time.milliTimestamp();
+        const data = try serializeRepairRequest(
+            self.allocator,
+            request.request,
+            self.keypair,
+            request.recipient,
+            @intCast(timestamp),
+            self.rng.int(Nonce),
+        );
+        defer self.allocator.free(data);
+        const addr = request.recipient_addr.toString();
+        self.logger.infof(
+            "sending repair request to {s} - {}",
+            .{ addr[0][0..addr[1]], request.request },
+        );
+        _ = try self.udp_send_socket.sendTo(request.recipient_addr.toEndpoint(), data);
+    }
+};
+
+/// A repair request plus its destination.
+pub const AddressedRepairRequest = struct {
+    request: RepairRequest,
+    recipient: Pubkey,
+    recipient_addr: SocketAddr,
+};
+
+/// How many slots to cache in RepairPeerProvider
+const REPAIR_PEERS_CACHE_CAPACITY: usize = 128;
+/// Maximum age of a cache item to use for repair peers
+const REPAIR_PEERS_CACHE_TTL_SECONDS: u64 = 10;
+
+/// A node that can service a repair request.
+pub const RepairPeer = struct {
+    pubkey: Pubkey,
+    serve_repair_socket: SocketAddr,
+};
+
+/// Provides peers that repair requests can be sent to.
+///
+/// TODO benchmark the performance of some alternate approaches, for example:
+/// - directly grab a single random peer from gossip instead
+///      of the entire list (good if we don't access a slot many times)
+/// - single sorted cache for all slots with a binary search to filter by slot
+///     - upside is fewer table locks
+///     - good if we're mainly looking at older slots, not the last few slots
+///     - downside is it may get stale and not represent the latest slots,
+///         unless cache TTL is reduced to less than 1 slot duration, in which
+///         case it may defeat the purpose of this approach.
+/// The key for these benchmarks is to understand the actual repair requests that
+/// are being requested on mainnet. There are trade-offs for different kinds
+/// of requests. Naive benchmarks will optimize the wrong behaviors.
+pub const RepairPeerProvider = struct {
+    allocator: Allocator,
+    rng: Random,
+    gossip: *RwMux(GossipTable),
```

Comment:
nit: can we change this field name (gossip -> gossip_table_rw) to be consistent

Reply:
all renames in 1 commit: https://github.com/Syndica/sig/pull/87/commits/4e02024cfbff9143db5fdf9ec08dbb32a2cef83a